### PR TITLE
Use stamping approach to deposit off-axis cells

### PR DESCRIPTION
**Pull request by cphyc**
## PR Summary

On main, off-axis projections are very slow due to the need to cast rays through the domain.
This is particularly the case for octree-datasets to to the hot-loop [for grid, mask in data_source.blocks:](https://github.com/cphyc/yt/blob/f8d38de9fc71cd83b40ab74eb2527644bbc1f4be/yt/visualization/volume_rendering/off_axis_projection.py#L446-L464). Indeed, this iterates over all octs in the domain, one-by-one.

While one solution might be to move this chunk to a Cython-accelerated code, it will remain somewhat slow no matter what.

In this PR, I take a completely different approach.
In the first part of the code, I compute the projected depth of a cube as seen from the camera position (i.e. how long is the segment traversing the cube at different pixel locations). One way of seeing this is that we now have a precomputed kernel, which we can apply to each cell at a scale proportional to the cell size.

This yields good performances: ~2s on an i9 for an 800x800 simulation with 10,000,000 cells. Interestingly, this could also be used to offload the rendering to the GPU.

While the code is relatively stable, I still get slight differences between axis-aligned projections and this method, which suggests that there are some minor details that I got wrong (see below).

## Comparisons
Using the code
```python
import yt

ds = yt.load_sample("output_00080")
p = yt.ProjectionPlot(
    ds,
    "y",
    "density",
    weight_field="density"
)
p.set_zlim(("gas", "density"), 6e-31, 2e-26)
p.save("/tmp/projection.png")
p = yt.OffAxisProjectionPlot(
    ds,
    [0, 1, 0],
    "density",
    north_vector=[1, 0, 0],
    width=(1, "unitary"),
    center=[0.5, 0.5, 0.5],
    buff_size=(800, 800),
    weight_field="density",
)
p.set_zlim(("gas", "density"), 6e-31, 2e-26)
p.set_axes_unit("Mpc")
p.save("/tmp/new.png")
```

I obtain (as of #f8d38de) very similar performances as the on-axis projection (!!!)
|  | Result from `ProjectionPlot` | Result from new off-axis projection |
|--|---|---|
| Image | projection | new |
| Timing (s) | 1.45 | 2.14 (8 cores), 3.13 (1 core) |

### CODE REVIEW

File: yt/utilities/lib/image_utilities.pyx [hunk 1]
```diff
@@ -36,6 +41,370 @@ def add_points_to_greyscale_image(
...
+
+    # If the cell is fully within one pixel
+    if (jmax == jmin + 1) and (kmax == kmin + 1):
+        buffer[jmin][kmin] += w * q * dx
+        buffer_weight[jmin][kmin] += w * dx
+        return
+
+    depth = int(ceil(log2(2 * sqrt(3) * dx * float(max(Nx, Ny)))))
+    if depth < 1: depth = 1
+    if depth >= max_depth: depth = max_depth - 1
```
Comment (matthewturk) [inline yt/utilities/lib/image_utilities.pyx:95]: could we use `iclip`?

File: yt/utilities/lib/image_utilities.pyx [hunk 2]
```diff
@@ -36,6 +41,370 @@ def add_points_to_greyscale_image(
...
+
+    jmax = min(Nsx, 1 << depth)
+    kmax = min(Nsy, 1 << depth)
```
Comment (matthewturk) [inline yt/utilities/lib/image_utilities.pyx:98]: `imin`?  I think this might call the python function `min`.

File: yt/utilities/lib/image_utilities.pyx [hunk 3]
```diff
@@ -36,6 +41,370 @@ def add_points_to_greyscale_image(
...
+    for j in range(jmax):
+        xx1 = ((j - jmax / 2.) * dx_loc + x) * Nx
+        xx2 = ((j + 1 - jmax / 2.) * dx_loc + x) * Nx
+
+        jj1 = int(xx1)
+        jj2 = int(xx2)
+        # The subcell is out of the projected area
+        if jj2 < 0 or jj1 >= Nx: continue
+
+        dvx = max(0, min(1, (jj2 - xx1) / (xx2 - xx1)))
```
Comment (matthewturk) [inline yt/utilities/lib/image_utilities.pyx:114]: `iclip` again

File: yt/utilities/lib/image_utilities.pyx [hunk 4]
```diff
@@ -36,6 +41,370 @@ def add_points_to_greyscale_image(
...
+
+    # with gil: print(jmax, kmax)
+
+    for j in range(jmax):
+        xx1 = ((j - jmax / 2.) * dx_loc + x) * Nx
+        xx2 = ((j + 1 - jmax / 2.) * dx_loc + x) * Nx
+
+        jj1 = int(xx1)
+        jj2 = int(xx2)
+        # The subcell is out of the projected area
```
Comment (matthewturk) [inline yt/utilities/lib/image_utilities.pyx:111]: Can we cast these instead of function calling?

File: yt/utilities/lib/image_utilities.pyx [hunk 5]
```diff
@@ -36,6 +41,370 @@ def add_points_to_greyscale_image(
...
+            # The subcell is out of the projected area
+            if kk2 < 0 or kk1 >= Ny: continue
+
+            # with gil:
+            #     print(f"{j=} {k=} {xx1=} {xx2=} {yy1=} {yy2=} {jj1=} {jj2=} {kk1=} {kk2=}")
+
+            sw =  stamp[depth, j, k] * w * dx
+            swq = sw * q
+
+            dvy = max(0, min(1, (kk2 - yy1) / (yy2 - yy1)))
```
Comment (matthewturk) [inline yt/utilities/lib/image_utilities.pyx:134]: `iclip`

File: yt/utilities/lib/image_utilities.pyx [hunk 6]
```diff
@@ -36,6 +41,370 @@ def add_points_to_greyscale_image(
...
+
+        for k in range(kmax):
+            if stamp_mask[depth, j, k] == 0:
+                continue
+
+            yy1 = ((k - kmax / 2.) * dy_loc + y) * Ny
+            yy2 = ((k + 1 - kmax / 2.) * dy_loc + y) * Ny
+
+            kk1 = int(yy1)
+            kk2 = int(yy2)
```
Comment (matthewturk) [inline yt/utilities/lib/image_utilities.pyx:124]: cast here too?

File: yt/utilities/lib/image_utilities.pyx [hunk 7]
```diff
@@ -36,6 +41,370 @@ def add_points_to_greyscale_image(
...
+    sq3 = sqrt(3.)
+
+    # Local buffers
+    cdef np.float64_t **lbuffer
+    cdef np.float64_t **lbuffer_weight
+
+    cdef int num_particles = len(Xp)
+
+    with nogil, parallel():
+        lbuffer = <np.float64_t**> malloc(sizeof(np.float64_t*) * Nx)
```
Comment (matthewturk) [inline yt/utilities/lib/image_utilities.pyx:228]: Is this a better solution than allocating the full `Nx*Ny` array?  If I am reading it right, this will still allocate the full thing on every processor, won't it?

File: yt/utilities/lib/image_utilities.pyx [hunk 8]
```diff
@@ -36,6 +41,370 @@ def add_points_to_greyscale_image(
...
+        lbuffer_weight = <np.float64_t**> malloc(sizeof(np.float64_t*) * Nx)
+        for j in range(Nx):
+            lbuffer[j] = <np.float64_t*> malloc(sizeof(np.float64_t) * Ny)
+            lbuffer_weight[j] = <np.float64_t*> malloc(sizeof(np.float64_t) * Ny)
+            for k in range(Ny):
+                lbuffer[j][k] = 0
+                lbuffer_weight[j][k] = 0
```
Comment (matthewturk) [inline yt/utilities/lib/image_utilities.pyx:235]: might be able to use `calloc` instead of this

File: yt/utilities/lib/image_utilities.pyx [hunk 9]
```diff
@@ -36,6 +41,370 @@ def add_points_to_greyscale_image(
...
+        # Free memory
+        for j in range(Nx):
+            free(lbuffer[j])
+            free(lbuffer_weight[j])
+        free(lbuffer)
+        free(lbuffer_weight)
+
+
+@cython.boundscheck(False)
+cdef inline np.float64_t det2d(const np.float64_t[::1] a, const np.float64_t[::1] b) noexcept nogil:
```
Comment (matthewturk) [inline yt/utilities/lib/image_utilities.pyx:286]: this might still do an initialization check, which can be expensive

File: yt/utilities/lib/image_utilities.pyx [hunk 10]
```diff
@@ -36,6 +41,370 @@ def add_points_to_greyscale_image(
...
+    return a[0] * b[1] - a[1] * b[0]
+
+@cython.cdivision(True)
+cdef bint check_in_parallelogram(
```
Comment (matthewturk) [inline yt/utilities/lib/image_utilities.pyx:290]: boundscheck and initializedcheck might add speed

File: yt/utilities/lib/image_utilities.pyx [hunk 11]
```diff
@@ -36,6 +41,370 @@ def add_points_to_greyscale_image(
...
+                zmax = max(z, zmax)
+                Nhit += 1
+
+            within = check_in_parallelogram(ObackA, w2d, u2d, -1, -1, nm)
+            if within:
+                z = Oback[2] + nm[0] * u[2] + nm[1] * w[2]
+                zmin = min(z, zmin)
+                zmax = max(z, zmax)
+                Nhit += 1
+
```
Comment (matthewturk) [inline yt/utilities/lib/image_utilities.pyx:399]: I think if we use the Cythonized `imin` and `imax` it might go a bit faster

File: yt/utilities/lib/image_utilities.pyx [hunk 12]
```diff
@@ -36,6 +41,370 @@ def add_points_to_greyscale_image(
...
+            if Nhit == 0:
+                continue
+            elif Nhit == 1:
+                raise RuntimeError("This should not happen")
+            else:
+                buffer[i, j] += zmax - zmin
+                buffer_mask[i, j] |= 1
```
Comment (matthewturk) [inline yt/utilities/lib/image_utilities.pyx:406]: Doesn't even need to be an or -- could just set it.

**Comment by chrishavlin** [comment]
No insight into why you're getting different results between the on-axis projection and the off-axis when aligned with the axis, but here's an example that results in very different behavior. The standard on axis projection returns what I would expect:

```python
fld = ("ramses", "z")
wt = ("index", "ones")
p = yt.ProjectionPlot(
    ds, 
    "y",
    fld,
    weight_field=wt,
    buff_size=(800,800),
)
p.set_axes_unit("Mpc")
p.set_unit(fld, "Mpc")
p.save('old.png')
```
old

But the new off axis produces a constant value of 1? 

```python
p = yt.OffAxisProjectionPlot(
    ds, 
    [0, 1, 0], 
    fld,
    north_vector=[1, 0, 0], 
    width=(1, "unitary"),
    center=[0.5, 0.5, 0.5],
    buff_size=(800, 800),
    weight_field=wt,
)
p.set_axes_unit("Mpc")
p.set_unit(fld, "Mpc")

mx = p.plots[fld].image.get_array().max()
mn = p.plots[fld].image.get_array().min()
print((mn, mx))
# (1.0, 1.0)
p.save('new.png')
```
new

Been reading and stepping through the code, I kinda think it's related to how the weighting field is handled? maybe units related?

File: yt/utilities/lib/image_utilities.pyx [hunk 13]
```diff
@@ -36,6 +41,370 @@ def add_points_to_greyscale_image(
...
+    const np.float64_t[::1] qty,
+    const np.float64_t[::1] weight,
+    const np.float64_t[:, :] rotation,
+    np.float64_t[:, ::1] buffer,
+    np.float64_t[:, ::1] buffer_weight,
+    const int Nx,
+    const int Ny,
+    const int Npix_min = 4,
+):
+    cdef np.ndarray[np.float64_t, ndim=1] center = np.array([0.5, 0.5, 0.5])
```
Comment (chrishavlin) [inline yt/utilities/lib/image_utilities.pyx:168]: why is this fixed rather than an argument?

Comment (cphyc) [reply yt/utilities/lib/image_utilities.pyx:168]: In the "stamping" approach, I compute the projection of a cube of size 1 centred on [0.5, 0.5, 0.5]. This is then "stamped" at the location of each cell. The centre here relates to the centre of this reference cube, not to where the centre actually is. This is done from the caller's side of things.

File: yt/utilities/lib/image_utilities.pyx (hunk 9 above)
Comment (cphyc) [reply yt/utilities/lib/image_utilities.pyx:292]: I just checked, and it does not seem to be checking anything
```
 290: @cython.boundscheck(False)
 291: @cython.wraparound(False)
+292: cdef inline np.float64_t det2d(const np.float64_t[::1] a, const np.float64_t[::1] b) noexcept nogil:
static CYTHON_INLINE __pyx_t_5numpy_float64_t __pyx_f_2yt_9utilities_3lib_15image_utilities_det2d(__Pyx_memviewslice __pyx_v_a, __Pyx_memviewslice __pyx_v_b) {
  __pyx_t_5numpy_float64_t __pyx_r;
/* … */
  /* function exit code */
  __pyx_L0:;
  return __pyx_r;
}
+293:     return a[0] * b[1] - a[1] * b[0]
  __pyx_t_1 = 0;
  __pyx_t_2 = 1;
  __pyx_t_3 = 1;
  __pyx_t_4 = 0;
  __pyx_r = (((*((__pyx_t_5numpy_float64_t const  *) ( /* dim=0 */ ((char *) (((__pyx_t_5numpy_float64_t const  *) __pyx_v_a.data) + __pyx_t_1)) ))) * (*((__pyx_t_5numpy_float64_t const  *) ( /* dim=0 */ ((char *) (((__pyx_t_5numpy_float64_t const  *) __pyx_v_b.data) + __pyx_t_2)) )))) - ((*((__pyx_t_5numpy_float64_t const  *) ( /* dim=0 */ ((char *) (((__pyx_t_5numpy_float64_t const  *) __pyx_v_a.data) + __pyx_t_3)) ))) * (*((__pyx_t_5numpy_float64_t const  *) ( /* dim=0 */ ((char *) (((__pyx_t_5numpy_float64_t const  *) __pyx_v_b.data) + __pyx_t_4)) )))));
  goto __pyx_L0;
```

File: yt/utilities/lib/image_utilities.pyx (hunk 10 above)
Comment (cphyc) [reply yt/utilities/lib/image_utilities.pyx:298]: done!

**Comment by cphyc** [comment]
I have *finally* updated the code. Here are some examples

# Example 1: One of the galaxy in `output_00080`
```python
import yt

ds = yt.load("output_00080")
ad = ds.all_data()
center = ds.arr(ad.argmax("density"))
width = ds.quan(200, "kpc")
zlim = (6e-30, 2e-24)
data_source = ds.sphere(center, width * 2)

t0 = time()
p = yt.ProjectionPlot(
    ds,
    "x",
    "density",
    buff_size=(800, 800),
    weight_field="density",
    center=center,
    width=width,
    data_source=data_source
)
pref = p
p.set_zlim(("gas", "density"), pref.frb["gas", "density"].min(), pref.frb["gas", "density"].max())
p.save("/tmp/projection.png")

p = yt.OffAxisProjectionPlot(
    ds,
    [1, 0, 0],
    "density",
    north_vector=[0, 0, 1],
    buff_size=(800, 800),
    weight_field="density",
    center=center,
    width=width,
    data_source=data_source
)
p.set_zlim(("gas", "density"), pref.frb["gas", "density"].min(), pref.frb["gas", "density"].max())
p.set_axes_unit("Mpc")
p.save("/tmp/new.png")
```
| On axis projection | Off axis projection | Off axis projection (from [1, 2, 3]) |
|----|----|---|
| on-axis projection | off-axis projection | off-axis projection, different view |


# Example 2: plot of the z component
This is @chrishavlin example from https://github.com/yt-project/yt/pull/4741#issuecomment-1813024283.
| On axis projection | Off axis projection |
|----|----|
| old | new |

File: yt/visualization/volume_rendering/off_axis_projection.py [hunk 14]
```diff
@@ -380,32 +383,90 @@ def temp_weightfield(field, data):
...
+            axis=-1,
+        )
+
+        for idim, periodic in enumerate(data_source.ds.periodicity):
+            if not periodic:
+                continue
+            # Wrap into [-0.5, +0.5]
+            xyz[..., idim] = (xyz[..., idim] + 0.5) % 1 - 0.5
+
+        dx = (data_source["index", "dx"] / wmax).to("1").d
```
Comment (chrishavlin) [inline yt/visualization/volume_rendering/off_axis_projection.py:412]: are octree grids always cubes? I naively expected to see some handling of `dy` and `dz` as well, but this all seems to assume dx=dy=dz.

**Review by chrishavlin — COMMENTED** [review]
Looks great! Some minor comments that you can take or leave, will hit approve after an answer to my question about grid spacing :)

File: yt/visualization/volume_rendering/off_axis_projection.py (hunk 14 above)
Comment (cphyc) [reply yt/visualization/volume_rendering/off_axis_projection.py:425]: I have added a comment in that sense. The octree grids are always cubic in the astronomical codes I know, but it might not be the general case. I've added a comment in that direction and an explicit check.

**Comment by neutrinoceros** [comment]
I would have liked to review this but unfortunately I'm approaching review request bankruptcy so realistically it's not going to happen. I think it would be fine to merge it with Chris' approval. For release notes though, should this be in new features, perf improvements, minor API change, all three ?

**Comment by cphyc** [comment]
I added labels (improvement and new feature). The API doesn't change, it's a drop-in replacement for the existing off-axis renderer.

**Comment by neutrinoceros** [comment]
Let me rephrase that: is there any observable change besides improved performance ?

**Comment by cphyc** [comment]
Yes; the method does not yield _exactly_ the same images as the previous implementation (although they should be extremely close).

**Comment by chrishavlin** [comment]
I didn't mention it in my review, but while reviewing i did actually run some more comparison tests between old and new methods with a couple datasets and some different cutting planes. the differences were on the order of 1e-40 relative difference in all the cases I looked at. it wasn't exhaustive testing by any means, but I think it's pretty safe to say that it's unlikely for folks to observe any difference by eye but may see differences when comparing actual pixel values. Certainly worth giving a heads up in the release notes though!

**Comment by neutrinoceros** [comment]
thank you both ! Feel free to merge away then !
